### ISSUE TITLE
Question: how to create `EnumSet` from enum constant with non-default annotation

### ISSUE THREAD

**Issue by msridhar** [typetools/checker-framework#6110]
My apologies if this is already covered in the manual or another issue; I couldn't find it.  Consider the following example:
```java
import org.checkerframework.checker.tainting.qual.*;
import java.util.EnumSet;
class Test {
  @SuppressWarnings("assignment") // #2156 
  static enum TestEnum { ONE, @Untainted TWO }
  static void test() {
      EnumSet<@Tainted TestEnum> s1 = EnumSet.of(TestEnum.ONE);
      EnumSet<@Untainted TestEnum> s2 = EnumSet.of(TestEnum.TWO);
  }
}
```
After suppressing the warning caused by #2156, we get two more warnings on the assignment to `s2`:
```
/tmp/Test.java:10: error: [type.argument] incompatible type argument for type parameter E of EnumSet.
      EnumSet<@Untainted TestEnum> s2 = EnumSet.of(TestEnum.TWO);
              ^
  found   : @Untainted TestEnum
  required: @Tainted Enum<@Untainted TestEnum>
/tmp/Test.java:10: error: [type.argument] incompatible type argument for type parameter E extends @Tainted Enum<@Untainted TestEnum> of EnumSet.of.
      EnumSet<@Untainted TestEnum> s2 = EnumSet.of(TestEnum.TWO);
                                                  ^
  found   : @Untainted TestEnum
  required: @Tainted Enum<@Untainted TestEnum>
2 errors
```
Is there some way to annotate the assignment to `s2` to make it pass the checker?

**Comment by smillst** [comment]
I can't figure out  a way to do this.  I think this is a bug.

**Comment by msridhar** [comment]
Thanks for looking into it!  If it's helpful, here is a possibly-related example that @nimakarimipour put together:
```java
import org.checkerframework.checker.tainting.qual.Untainted;
class Test {
  public void test(@Untainted Foo foo) {
    get(foo);
  }
  static <E extends Bar<E>> E get(E e) {
    return null;
  }
  static class Bar<E extends Bar<E>> {
  }
  static class Foo extends Bar<@Untainted Foo> { }    
}
```
This version passes the Tainting Checker.  But if we write `static class Foo extends Bar<Foo> { }` instead (without the `@Untainted` qualifier on the type parameter) we get an error:
```
/tmp/Test.java:4: error: [type.argument] incompatible type argument for type parameter E extends @Tainted Bar<@Untainted Foo> of Test.get.
    get(foo);
       ^
  found   : @Untainted Foo
  required: @Tainted Bar<@Untainted Foo>
```
This may be related since the `java.lang.Enum` class is declared as `class Enum<E extends Enum<E>>`, similar to how the `Bar` class above is declared.

**Comment by smillst** [comment]
Yes that's the problem.  We actually have special casing for enums that is trying to to fix this:
https://github.com/typetools/checker-framework/blob/d607f262c4ac8039785674e7feb46c3e4c16fb70/framework/src/main/java/org/checkerframework/framework/util/AnnotatedTypes.java#L149-L155

But it doesn't really make sense. Nor does it actually work.

I can change it so that the super type of `@Untainted TestEnum` is `@Untainted Enum<@Untainted TestEnum>`, but then I get errors when using `Enum#compareTo`:

```
import org.checkerframework.checker.tainting.qual.Untainted;

class Issue6110 {
  @SuppressWarnings("assignment") // #2156
  enum TestEnum {
    ONE,
    @Untainted TWO
  }

  static void test(Enum<@Untainted TestEnum> o) {
    o.compareTo(TestEnum.ONE);
  }
}
``` 

```
../testcases/src/Issue6110.java:11: error: [argument] incompatible argument for parameter o of Enum.compareTo.
    o.compareTo(TestEnum.ONE);
                        ^
  found   : @Tainted TestEnum
  required: @Untainted TestEnum
1 error
```

I can fix this error by annotation `compareTo` in the annotated JDK like this:
```
int compareTo(@Tainted E o)
```

But overall I'm not sure if these changes are sound.

**Comment by msridhar** [comment]
Thanks a lot for this.  I'm trying to understand things step by step.  So for this error with the example of `Foo` and `Bar`:
```
/tmp/Test.java:4: error: [type.argument] incompatible type argument for type parameter E extends @Tainted Bar<@Untainted Foo> of Test.get.
    get(foo);
       ^
  found   : @Untainted Foo
  required: @Tainted Bar<@Untainted Foo>
```
I thought the issue here was that `Foo` is always a subtype of `Bar<@Tainted Foo>` but not always a subtype of `Bar<@Untainted Foo>`.  (It would be cool if we could somehow express that `@Tainted Foo` is a subtype of `Bar<@Tainted Foo>` and that `@Untainted Foo` is a subtype of `Bar<@Untainted Foo>`, but I don't think that is possible?)  I thought that adding a `@Covariant` annotation might be an alternate way to fix this:
```java
class Test {
  public void test(@Untainted Foo foo) {
      get(foo);
  }
  static <E extends Bar<E>> E get(E e) {
    return e;
  }
  @Covariant(0)  
  static class Bar<E extends Bar<E>> {
  }
  static class Foo extends Bar<Foo> { }    
}
```
I expected this would make `@Untainted Foo` a subtype of `@Tainted Bar<@Untainted Foo>` but it seemed to have no effect; maybe I misunderstood something?  I was hoping some kind of similar `@Covariant` solution might work for enums, since they are almost always immutable.

**Comment by smillst** [comment]
> ```
> /tmp/Test.java:4: error: [type.argument] incompatible type argument for type parameter E extends @Tainted Bar<@Untainted Foo> of Test.get.
>     get(foo);
>        ^
>   found   : @Untainted Foo
>   required: @Tainted Bar<@Untainted Foo>
> ```
> 
> I thought the issue here was that `Foo` is always a subtype of `Bar<@Tainted Foo>` but not always a subtype of `Bar<@Untainted Foo>`. 

Right, `@Untainted Foo` is a subtype of  `@Untainted Bar<@Tainted Foo>` and `@Tainted Foo` is a subtype of  `@Tainted Bar<@Tainted Foo>`.
Neither  `@Untainted Foo` nor  `@Tainted Foo` is a subtype of `@Taintined Bar<@Untainted Foo>`.

> (It would be cool if we could somehow express that `@Tainted Foo` is a subtype of `Bar<@Tainted Foo>` and that `@Untainted Foo` is a subtype of `Bar<@Untainted Foo>`, but I don't think that is possible?)

I think this is what the special casing in the framework was trying, but failing to do. This is also breaking subtying rules, so we may run into soundness issues. 

> I thought that adding a `@Covariant` annotation might be an alternate way to fix this:
> 
> ```java
> class Test {
>   public void test(@Untainted Foo foo) {
>       get(foo);
>   }
>   static <E extends Bar<E>> E get(E e) {
>     return e;
>   }
>   @Covariant(0)  
>   static class Bar<E extends Bar<E>> {
>   }
>   static class Foo extends Bar<Foo> { }    
> }
> ```
> 
> I expected this would make `@Untainted Foo` a subtype of `@Tainted Bar<@Untainted Foo>` but it seemed to have no effect; maybe I misunderstood something? I was hoping some kind of similar `@Covariant` solution might work for enums, since they are almost always immutable.

`@Covariant` makes  `@Tainted Bar<@Untainted Foo>`  a subtype of `@Tainted Bar<@Tainted Foo>`.  But `@Untainted Foo` is a direct subtype of `@Untainted Bar<@Tainted Foo>` which even using `@Covariant` is still not a subtype of `@Untainted Bar<@Untainted Foo>`.

**Comment by smillst** [comment]
What we want is probably closer to writing `@PolyTainted` on `Foo` in the extends clause of `Foo` (This isn't implemented nor designed):

```java
 class Foo extends Bar<@PolyTainted Foo> { }    
```
which would mean that the direct super type of `@Untainted Foo` is `@Untainted Bar<@Untainted Foo>` and `@Tainted Foo` is `@Tainted Bar<@Tainted Foo>`.  But there's a problem: 

1. `@Untainted Foo <:  @Tainted Foo`
2. `@Untainted Foo <: @Untainted Bar<@Untainted Foo>` 
3. `@Tainted Foo <: @Tainted Bar<@Tainted Foo>`

The above three statements imply that `@Untainted Bar<@Untainted Foo> <: @Tainted Bar<@Tainted Foo>` but that's only true if Bar is covariant.

We would need to think carefully about the right type rules before we try to implement this.

**Comment by msridhar** [comment]
Thanks a lot for the investigation and explanations, @smillst.  I think the scenario we care about most is for enums rather than the `Foo` / `Bar` example.  I think it would be reasonable to assume that most (nearly all?) enum values are immutable, in which case covariance is safe I think.  Further, in our actual use case, we'd like to treat _all_ enum constants as `@Untainted` by default, so the `compareTo` issue you pointed out would not arise.  Probably this overall issue needs a bit more thought and also feedback from @mernst.  In the meantime, could you possibly share the change you prototyped that makes `@Untainted Enum<@Untainted TestEnum>` the supertype of `@Untainted TestEnum`?  We could test that change out for our checker variant to see if we run into other weird behaviors.  Thanks again!

**Comment by smillst** [comment]
Here's the branch with special casing for enums:
https://github.com/smillst/checker-framework/tree/issue6110

>  we'd like to treat all enum constants as @Untainted by default

In this case, you could  annotate Enum in the jdk:
```
@DefaultQualifierForUse(Untainted.class)
 public abstract class Enum<E extends @Tainted Enum<E>>
```

Or use the meta-annotation `@DefaultFor(types=Enum.class)` on the declaration of `Untainted`.  (You will still need to add `@Tainted` to `class Enum<E extends @Tainted Enum<E>>`.)

Or if enum constants should only be `@Untainted` and never `@Tainted`, then 
```
@Untainted
 public abstract class Enum<E extends Enum<E>>
```

**Comment by msridhar** [comment]
Thanks a lot @smillst!

**Comment by msridhar** [comment]
As an update, the changes at https://github.com/smillst/checker-framework/tree/issue6110 work perfectly for our checker and use case (all unexpected errors with enums are eliminated).

**Comment by txiang61** [comment]
Hello I'm have a similar issue with the signedness checker.

```
error: [type.argument.type.incompatible] incompatible type argument for type parameter E extends @Signed Enum<@Signed EvaluationResponseFlags> of noneOf.
      return EnumSet.noneOf(EvaluationResponseFlags.class);
                            ^
  found   : @Signed EvaluationResponseFlags
  required: @Signed Enum<@Signed EvaluationResponseFlags>
```
Here is another one which I think is also a bug
```
error: [type.argument.type.incompatible] incompatible type argument for type parameter E of EnumSet.
  public EnumSet<EvaluationResponseFlags> getFlags() {
                 ^
  found   : @Signed EvaluationResponseFlags
  required: @UnknownSignedness Enum<@Signed EvaluationResponseFlags>
```
I had these suppressed for now.
